modules/tensor_mechanics/src/userobjects/TensorMechanicsPlasticTensileMulti.C

```diff
@@ -129,111 +135,528 @@ TensorMechanicsPlasticTensileMulti::activeConstraints(const std::vector<Real> &
+  // trial_order[1] = type of return to try second
+  // trial_order[2] = type of return to try third
+  std::vector<int> trial_order(3);
+  if (yf[0] > 0) // all the yield functions are positive, since eigvals are ordered eigvals[0] <= eigvals[1] <= eigvals[2]
   {
-    act[0] = act[1] = act[2] = true;
-    returned_stress(0, 0) = returned_stress(1, 1) = returned_stress(2, 2) = str;
-    returned_stress = eigvecs*returned_stress*(eigvecs.transpose());
-    return;
+    trial_order[0] = tip;
+    trial_order[1] = edge;
+    trial_order[2] = plane;
+  }
+  else if (yf[1] > 0)  // two yield functions are positive
+  {
+    trial_order[0] = edge;
+    trial_order[1] = tip;
+    trial_order[2] = plane;
+  }
+  else
+  {
+    trial_order[0] = plane;
+    trial_order[1] = edge;
+    trial_order[2] = tip;
   }
 
-  // 011 (edge)
-  std::vector<Real> n1xn2(3);
-  n1xn2[0] = n[1][1]*n[2][2] - n[1][2]*n[2][1];
-  n1xn2[1] = n[1][2]*n[2][0] - n[1][0]*n[2][2];
-  n1xn2[2] = n[1][0]*n[2][1] - n[1][1]*n[2][0];
-  // work out the point to which we would return, "a".  It is defined by
-  // f1 = 0 = f2, and that (p - a).(n1 x n2) = 0, where "p" is the
-  // starting position (p = eigvals).
-  // In the following a = (a0, str, str)
-  Real pdotn1xn2 = dot(eigvals, n1xn2);
-  Real a0 = (-str*n1xn2[1] - str*n1xn2[2] + pdotn1xn2)/n1xn2[0];
-  // we need p - a = alpha*n1 + beta*n2, where alpha and beta are non-negative
-  // for Kuhn-Tucker to be satisfied
-  std::vector<Real> pminusa(3);
-  pminusa[0] = eigvals[0] - a0;
-  pminusa[1] = v[1];
-  pminusa[2] = v[2];
-  if ((pminusa[2] - pminusa[0])/(1.0 - ratio) >= 0 && (pminusa[1] - pminusa[0])/(1.0 - ratio) >= 0)
+  unsigned trial;
+  bool nr_converged;
+  for (trial = 0 ; trial < 3 ; ++trial)
   {
-    returned_stress(0, 0) = a0;
-    returned_stress(1, 1) = str;
-    returned_stress(2, 2) = str;
-    returned_stress = eigvecs*returned_stress*(eigvecs.transpose());
-    act[1] = act[2] = true;
-    return;
+    switch (trial_order[trial])
+    {
+      case tip:
+        nr_converged = returnTip(eigvals, n, dpm, returned_stress, intnl_old, 0);
+        break;
+      case edge:
+        nr_converged = returnEdge(eigvals, n, dpm, returned_stress, intnl_old, 0);
+        break;
+      case plane:
+        nr_converged = returnPlane(eigvals, n, dpm, returned_stress, intnl_old, 0);
+        break;
+    }
+    str = tensile_strength(intnl_old + dpm[0] + dpm[1] + dpm[2]);
+    if (nr_converged && KuhnTuckerOK(returned_stress, dpm, str))
+      break;
   }
 
-  // 001 (plane)
-  // the returned point, "a", is defined by f2=0 and
-  // a = p - alpha*n2
-  Real alpha = (eigvals[2] - str)/n[2][2];
-  act[2] = true;
-  returned_stress(0, 0) = eigvals[0] - alpha*n[2][0];
-  returned_stress(1, 1) = eigvals[1] - alpha*n[2][1];
-  returned_stress(2, 2) = str;
+  if (trial == 3)
+  {
+    Moose::out << "Trial stress = \n";
+    trial_stress.print();
+    Moose::out << "Internal parameter = " << intnl_old << "\n";
```

Comment:
These `Moose::out` streams should probably be `Moose::err`. The reason for this becomes apparent when I split the streams which is often done on batch systems. Part of your error message would be in the standard output stream while the final error message would be in a different stream. 

To take this even further, your `print()` function on the stress should take an optional stream so that you can pass `Moose:err` to it in this context but have it print to stdout in the normal case.